utils/test_results.py

```diff
@@ -76,131 +298,152 @@ def generate_test_results(
+                    )
+                    .order_by("-runtime")
+                    .values("test_id")[0 : slow_test_threshold(num_tests)]
+                )
+                slowest_test_id_set = {test["test_id"] for test in slowest_test_ids}
 
-            totals = totals.filter(test_id__in=slowest_test_ids)
+                test_ids = (
+                    test_ids & slowest_test_id_set if test_ids else slowest_test_id_set
+                )
 
-    commits_where_fail_sq = (
-        totals.filter(test_id=OuterRef("test_id"))
-        .annotate(v=Distinct(Unnest(F("commits_where_fail"))))
-        .values("v")
+    if not first and not last:
+        first = 20
+
+    should_reverse = False if first else True
+
+    query = generate_base_query(
+        repoid=repoid,
+        ordering=ordering,
+        ordering_direction=ordering_direction,
+        should_reverse=should_reverse,
+        branch=branch,
+        interval_num_days=interval_num_days,
+        testsuites=testsuites,
+        term=term,
+        test_ids=test_ids,
     )
 
-    # TODO: add back in latest duration when performance is acceptable
-    #  latest_duration_sq = (
-    #     totals.filter(test_id=OuterRef("test_id"))
-    #     .values("last_duration_seconds")
-    #     .order_by("-latest_run")[:1]
-    # )
-
-    aggregation_of_test_results = totals.values("test").annotate(
-        total_test_count=Cast(
-            Sum(F("pass_count")),
-            output_field=FloatField(),
+    with connection.cursor() as cursor:
+        cursor.execute(
+            query.query,
+            query.params,
         )
-        + Cast(Sum(F("fail_count")), output_field=FloatField()),
-        total_fail_count=Cast(Sum(F("fail_count")), output_field=FloatField()),
-        total_flaky_fail_count=Cast(
-            Sum(F("flaky_fail_count")), output_field=FloatField()
-        ),
-        total_skip_count=Cast(Sum(F("skip_count")), output_field=FloatField()),
-        total_pass_count=Cast(Sum(F("pass_count")), output_field=FloatField()),
-        failure_rate=Case(
-            When(
-                total_test_count=0,
-                then=Value(0.0),
-            ),
-            default=F("total_fail_count") / F("total_test_count"),
-        ),
-        flake_rate=Case(
-            When(
-                total_test_count=0,
-                then=Value(0.0),
-            ),
-            default=F("total_flaky_fail_count") / F("total_test_count"),
-        ),
-        updated_at=Max("latest_run"),
-        commits_where_fail=ArrayLength(Array(Subquery(commits_where_fail_sq))),
-        last_duration=Value(0.0),
-        avg_duration=Avg("avg_duration_seconds"),
-        name=F("test__name"),
-        computed_name=F("test__computed_name"),
-    )
+        aggregation_of_test_results = cursor.fetchall()
+
+        rows = [TestResultsRow(*row) for row in aggregation_of_test_results]
+
+    page_size: int = first or last or 20
```

Comment:
at this point first or last will always have a value thanks to your check on 399 right?